### PR TITLE
use a common locking mechanism (filestream) for all platforms while writing to settings file

### PR DESCRIPTION
**Pull request by rohit21agrawal**
This is the correct solution to the issue : https://github.com/NuGet/Home/issues/2860

This makes use of a synchronized version of file locking by acquiring a filestream handle on a lock file. This approach works across all platforms and can do inter-process synchronization too.

CC: @joelverhagen @yishaigalatzer @alpaix @rrelyea 

(Once approved, will revert https://github.com/NuGet/NuGet.Client/commit/584723b5b378c2189f774dab097a4f295815b957 checked in via pull request : https://github.com/NuGet/NuGet.Client/pull/720

### CODE REVIEW

File: src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs [hunk 1]
```diff
@@ -176,5 +244,8 @@ private static char ToHexChar(int input)
                 return (char)(input + 0x30);
             }
         }
+
+        private const int NumberOfRetries = 3000;
```
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs]: nit: constants at the top

File: src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs [hunk 2]
```diff
@@ -176,5 +244,8 @@ private static char ToHexChar(int input)
...
+        private const int SleepDuration = 10;
```
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs]: nit: use `TimeSpan` to be explicit on units

File: src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs [hunk 3]
```diff
@@ -24,7 +24,7 @@ public static class ConcurrencyUtilities
             }
 
             // limit the number of unauthorized, this should be around 30 seconds.
-            var unauthorizedAttemptsLeft = 3000;
```
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs:27]: There's a test I think to verify the correctness of the async locking mechanism. The same test should be executed on the sync version.

**Comment by joelverhagen** [comment]
:watch:

File: src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs (hunk 2 above)
Comment (alpaix) [inline src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs]:
``` c#
TimeSpan.FromMilliseconds(10);
```

**Comment by alpaix** [comment]
Please add a test the locking mechanism works, IE it is able to wait and quit the while loop.
:watch:

**Comment by rohit21agrawal** [comment]
@joelverhagen @alpaix incorporated your feedback and added a test.

File: test/NuGet.Core.Tests/NuGet.Common.Test/ConcurrencyUtilitiesTests.cs [hunk 4]
```diff
@@ -12,6 +12,66 @@ namespace NuGet.Common.Test
...
+                const int threads = 50;
+                const int iterations = 10;
+
+                // This is the action that is execute inside of the lock.
+                Action lockedActionSync = () =>
+                {
+                    var acquired = verificationSemaphore.Wait(0);
+                    if (!acquired)
+                    {
+                        throw new InvalidOperationException();
```
Comment (alpaix) [inline test/NuGet.Core.Tests/NuGet.Common.Test/ConcurrencyUtilitiesTests.cs]: replace with `Assert.True` with explanatory error message

File: test/NuGet.Core.Tests/NuGet.Common.Test/ConcurrencyUtilitiesTests.cs [hunk 5]
```diff
@@ -12,6 +12,66 @@ namespace NuGet.Common.Test
...
+                    }
+
+                    // Hold the lock for a little bit.
+                    Thread.Sleep(TimeSpan.FromMilliseconds(1));
+
+                    verificationSemaphore.Release();
+                    return;
```
Comment (alpaix) [inline test/NuGet.Core.Tests/NuGet.Common.Test/ConcurrencyUtilitiesTests.cs]: redundant?

Comment (rohit21agrawal) [inline test/NuGet.Core.Tests/NuGet.Common.Test/ConcurrencyUtilitiesTests.cs]: not sure what you mean?

Comment (alpaix) [inline test/NuGet.Core.Tests/NuGet.Common.Test/ConcurrencyUtilitiesTests.cs]: is `return` here needed?

Comment (rohit21agrawal) [inline test/NuGet.Core.Tests/NuGet.Common.Test/ConcurrencyUtilitiesTests.cs]: no, it's not. fixed.

**Comment by joelverhagen** [comment]
:shipit:

**Comment by alpaix** [comment]
Looks great! :shipit:

**Comment by rohit21agrawal** [comment]
Thanks... I have tested this codepath on Mono on Linux, .NET Framework and CoreCLR and haven't noticed any issues..so going ahead and checking this in.

**Comment by rohit21agrawal** [comment]
Merged : https://github.com/NuGet/NuGet.Client/commit/f60773877d147b686411576915ef9e4d88d6fddb

**Pull request closed without merging**